### PR TITLE
User Unit Tests

### PR DESCRIPTION
**Pull request by 4rt3ry**
## Description 
Added unit tests for the User model

## Updates
Unit tests previously checked the various service outputs. Now they only check the controller outputs (mostly either returns a thing or throws a thing)

####  Create User (Controller)
    √ Creating a user should return the new user (11 ms)
    √ Creating a user with bad data should return an Error object (3 ms)
####  Authenticate User (Controller)
    √ Authenticating should return the user or undefined (1 ms)
    √ Authenticating user with bad data should return an Error object (1 ms)
####  Change Password (Controller)
    √ Changing password should return a boolean for success (2 ms)
    √ Changing password to the same one should return an Error object (2 ms)
    √ Changing password with bad data should return an Error object (1 ms)
####  Change Display Name (Controller)
    √ Changing display name should return a boolean for success (1 ms)
    √ Changing display name with bad data should return an Error object (2 ms)
####  Get User By ID (Controller)
    √ Getting user should return either a user object or undefined (1 ms)
    √ Get user using bad data should return an Error object (1 ms)

### CODE REVIEW

File: src/tests/user.test.ts [hunk 1]
```diff
@@ -1,44 +1,166 @@
...
+        (userService.createUser as jest.Mock).mockReturnValue(() => { throw new ValidationError('Invalid display name', []); });
+        const newUser = await _createUserController(sequelizeMock())(user.badEmail, user.password, '');
+        expect(newUser).toBeInstanceOf(ValidationError);
+    });
 
-        expect(response).toEqual(userData);
+    test('Create user with long display name should throw a validation error', async () => {
+        (userService.createUser as jest.Mock).mockReturnValue(() => { throw new ValidationError('Invalid display name', []); });
+        const newUser = await _createUserController(sequelizeMock())(user.badEmail, user.password, '>30 chars aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa');
+        expect(newUser).toBeInstanceOf(ValidationError);
```
Comment (BlckHawker) [inline src/tests/user.test.ts]: These tests seem unnecessary. The invalid inputs are not the reason why the `ValidationErrors` are being thrown, but rather because you are telling it to. There should only be two tests for this method:
1. returning the correct data
2. catching and returning a thrown error

File: src/tests/user.test.ts [hunk 2]
```diff
@@ -1,44 +1,166 @@
...
+        const authUser = await _authenticateController(sequelizeMock())(user.email, user.password);
+        expect(authUser).toEqual(userResponse);
+    });
 
-        const response = await _getUserByIDController(sequelizeMock)('8c2d50d1-9b1f-483c-b267-cecb929ffb97');
+    test('Authenticating non-existing user should return undefined', async () => {
+        (userService.authenticate as jest.Mock).mockReturnValue(() => Promise.resolve(undefined));
+        const authUser = await _authenticateController(sequelizeMock())(undefinedUser.email, undefinedUser.password);
+        expect(authUser).toBe(undefined);
+    });
```
Comment (BlckHawker) [inline src/tests/user.test.ts]: This test is unnecessary as it's the same as the one above. We are not testing specifically if a user exists or not as that's not the controller's responsibility.

File: src/tests/user.test.ts [hunk 3]
```diff
@@ -1,44 +1,166 @@
...
+        const passwordChange = await _changePasswordController(sequelizeMock())(user.email, user.password, user.newPassword);
+        expect(passwordChange).toBe(true);
+    });
+
+    test('Changing password with invalid authentication should return false', async () => {
+
+        // return doesn't matter since error is thrown before changePassword() is ever called
+        (userService.changePassword as jest.Mock).mockReturnValue(() => Promise.resolve(false));
+        const passwordChange = await _changePasswordController(sequelizeMock())(undefinedUser.email, undefinedUser.password, user.newPassword);
+        expect(passwordChange).toBe(false);
```
Comment (BlckHawker) [inline src/tests/user.test.ts]: These tests are duplicates

File: src/tests/user.test.ts [hunk 4]
```diff
@@ -1,44 +1,166 @@
...
 
-        const response = await _getUserByIDController(sequelizeMock)('8c2d50d1-9b1f-483c-b267-cecb929ffb97');
+    test('Authenticating non-existing user should return undefined', async () => {
+        (userService.authenticate as jest.Mock).mockReturnValue(() => Promise.resolve(undefined));
+        const authUser = await _authenticateController(sequelizeMock())(undefinedUser.email, undefinedUser.password);
+        expect(authUser).toBe(undefined);
+    });
 
-        expect(response).toBeUndefined();
+    test('Authenticating user with bad data should throw an error', async () => {
```
Comment (BlckHawker) [inline src/tests/user.test.ts]: This test description is not related to the test itself. This description would be good if we were testing `userService.authenticate` as it's that method's responsibility throw an error, not the controller's. The controller's responsibility to return the value of `userService.authenticate`, or an error it caught

File: src/tests/user.test.ts [hunk 5]
```diff
@@ -1,44 +1,166 @@
...
+        // return doesn't matter since error is thrown before changePassword() is ever called
+        (userService.changePassword as jest.Mock).mockReturnValue(() => Promise.resolve('can be literally anything'));
+        const passwordChange = await _changePasswordController(sequelizeMock())(user.email, user.password, user.password);
+        expect(passwordChange).toBeInstanceOf(Error);
+    });
+
+    test('Changing password with bad data should throw an error', async () => {
+
+        // return doesn't matter since error is thrown before changePassword() is ever called
+        (userService.changePassword as jest.Mock).mockReturnValue(() => { throw new Error(); });
```
Comment (BlckHawker) [inline src/tests/user.test.ts]: The mock can be removed in the scenario

File: src/tests/user.test.ts [hunk 6]
```diff
@@ -1,44 +1,166 @@
...
+        (userService.changeDisplayName as jest.Mock).mockReturnValue(() => Promise.resolve(true));
+        const displayNameChange = await _changeDisplayNameController(sequelizeMock())(user.email, user.password, user.newDisplayName);
+        expect(displayNameChange).toBe(true);
+    });
+
+    test('Changing display name with invalid authentication should return false', async () => {
+        (userService.changeDisplayName as jest.Mock).mockReturnValue(() => Promise.resolve(false));
+        const displayNameChange = await _changeDisplayNameController(sequelizeMock())(undefinedUser.email, undefinedUser.password, user.newDisplayName);
+        expect(displayNameChange).toBe(false);
+    });
```
Comment (BlckHawker) [inline src/tests/user.test.ts]: duplicate tests

File: src/tests/user.test.ts [hunk 7]
```diff
@@ -1,44 +1,166 @@
...
+
+    test('Changing display name with bad data should throw an error', async () => {
```
Comment (BlckHawker) [inline src/tests/user.test.ts]: inaccurate test description

File: src/tests/user.test.ts [hunk 8]
```diff
@@ -1,44 +1,166 @@
...
+        (userService.getUserByID as jest.Mock).mockReturnValue(() => Promise.resolve(userResponse));
+        const response = await _getUserByIDController(sequelizeMock())(user.id);
+        expect(response).toEqual(userResponse);
+    });
+
+    test('If the user does not exist, it should return undefined', async () => {
+        (userService.getUserByID as jest.Mock).mockReturnValue(() => Promise.resolve(undefined));
+        const response = await _getUserByIDController(sequelizeMock())(undefinedUser.id);
+        expect(response).toBeUndefined();
+    });
```
Comment (BlckHawker) [inline src/tests/user.test.ts]: duplicate tests

File: src/tests/user.test.ts [hunk 9]
```diff
@@ -1,44 +1,166 @@
...
 
-        expect(response).toBe(error);
+    test('Using bad data should throw an error', async () => {
```
Comment (BlckHawker) [inline src/tests/user.test.ts]: inaccurate test description

**Review by BlckHawker — CHANGES_REQUESTED** [review]
Good job on being thorough with your tests. 

- There are a lot of the descriptions that are inaccurate as they aren't related to the controller's responsibilities. I wrote a detailed description in one of the comments. I didn't want to make note of all of them as they would probably be overwhelming comment-wise.
- I made a note [here](https://github.com/RIT-Crowd-Comic/crowd-comic-back-end/pull/12#discussion_r1657131661) what I'm changing my naming convention to.
- There seems to be a lot of duplicate tests. Reminder to only test what the controller itself is doing. If you don't have any validators in a controller, it should probably only have 2 tests: one for returning whatever it's service method gave, and one for catching thrown errors.

File: src/tests/user.test.ts (hunk 1 above)
Comment (4rt3ry) [reply src/tests/user.test.ts]: The validation error happens within the service function, since the sequelize model itself has validators. I thought checking this mattered since the error was being thrown from the service function rather than the controller, but if all we're doing is checking the controller as if it was a black box, I can remove the test

**Review by BlckHawker — APPROVED** [review]
sonic-boom-knuckles
